### ISSUE TITLE
PhysgunPickup hook doesn't work properly on the client.

### ISSUE THREAD

**Issue by ARitz-Cracker** [Facepunch/garrysmod-issues#2279]
http://www.aritzcracker.ca/uploads/aritz/Garry's%20Mod%202015-11-01%2011_13_49%20AM.mp4

As you can see, use the PhysgunPickup hook to disable the my stencils on my entity, otherwise, the screen will flash.
However, if I put my mouse on the world, and then the entity falls onto my beam, the hook doesn't get called, and in extension, my screen flashes.

The hook gets called on the server, but not on the client.

**Comment by immibis** [comment]
Just FYI, you should be able to make that slot machine without stencils, if it's a major problem.

**Comment by mcd1992** [comment]
Have you reset the stencil masks prior to your stencil calls?

https://facepunch.com/showthread.php?t=1280421&p=41123348&viewfull=1#post41123348

**Comment by ARitz-Cracker** [comment]
Yes, I did. Also, Rubat made this function halo.RenderedEntity(). If it returns my entity, then don't draw the stencils. I only used my example the way because that was the code I used to discover the hook not working at the time.

**Comment by ARitz-Cracker** [comment]
Here's some code you can use
```
hook.Add( "PhysgunPickup", "asdfghjkl;'", function( ply, ent ) 
    if IsValid(ent) then
        notification.AddLegacy( ent:GetClass(), NOTIFY_GENERIC, 2 )
    end
end)
```
Just throw an entity in the air, and let it drop onto your physgun beam, you'll see that there won't be a notification when it crosses your beam.

**Comment by thegrb93** [comment]
I just tried it on my dedicated server with lua_run_cl and it worked for me.

**Comment by thegrb93** [comment]
Maybe you have a hook returning true somewhere.

**Comment by ARitz-Cracker** [comment]
As I said before, you must throw the entity into the air, and have it land on physgun beam while you have your fire button down. I tested this on a local listen server with minimal addons. (Only mine, which don't use that hook on the client)

**Comment by thegrb93** [comment]
Okay so it seems that on the client, each frame you are moving your mouse while aiming at the prop (before the server registers you are holding the prop) it will call the hook. On my server I can get 3 hook calls with just one pickup before the prop is 'held'. But if you start holding the prop and you aren't moving your mouse, the hook won't call (The hook will only call if you are moving your mouse).

**Comment by Jvs34** [comment]
Yeah that's called prediction, garry simply has the pickup code shared but the final say of course is only on the server, so the client still fires the hook anyway.

It's the same way with the gravity gun really, sometimes it will fire the punt hook clientside even though it was prevented serverside or it simply moved away.

I think the only way to fix this is to have the hook called on the client when the networked variable ( m_Hholdingent or something ) gets finally set by the server instead of the hook being predicted in the usual way.

**Comment by thegrb93** [comment]
No.. then we'll have an annoying delay of the clientside effects. I think the hook should be called every tick the client if aiming at an entity and trying to pick it up, rather than just when the client's mouse is moving.

**Comment by Kefta** [comment]
You shouldn't be relying on any predicted hook for deterministic client behaviour. Network to the client from the hook serverside if you want accurate results.
